diff --git a/chrome/installer/mini_installer/mini_file.cc b/chrome/installer/mini_installer/mini_file.cc
index 2e04f685379fb..118828250579f 100644
--- a/chrome/installer/mini_installer/mini_file.cc
+++ b/chrome/installer/mini_installer/mini_file.cc
@@ -6,6 +6,85 @@
 
 #include <utility>
 
+#pragma comment(lib, "ntdll.lib")
+
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
+typedef LONG NTSTATUS;
+
+typedef struct _IO_STATUS_BLOCK {
+    union {
+        NTSTATUS Status;
+        PVOID Pointer;
+    };
+
+    ULONG_PTR Information;
+} IO_STATUS_BLOCK, *PIO_STATUS_BLOCK;
+
+typedef enum _FILE_INFORMATION_CLASS {
+// end_wdm
+    FileDirectoryInformation         = 1,
+    FileFullDirectoryInformation,   // 2
+    FileBothDirectoryInformation,   // 3
+    FileBasicInformation,           // 4  wdm
+    FileStandardInformation,        // 5  wdm
+    FileInternalInformation,        // 6
+    FileEaInformation,              // 7
+    FileAccessInformation,          // 8
+    FileNameInformation,            // 9
+    FileRenameInformation,          // 10
+    FileLinkInformation,            // 11
+    FileNamesInformation,           // 12
+    FileDispositionInformation,     // 13
+    FilePositionInformation,        // 14 wdm
+    FileFullEaInformation,          // 15
+    FileModeInformation,            // 16
+    FileAlignmentInformation,       // 17
+    FileAllInformation,             // 18
+    FileAllocationInformation,      // 19
+    FileEndOfFileInformation,       // 20 wdm
+    FileAlternateNameInformation,   // 21
+    FileStreamInformation,          // 22
+    FilePipeInformation,            // 23
+    FilePipeLocalInformation,       // 24
+    FilePipeRemoteInformation,      // 25
+    FileMailslotQueryInformation,   // 26
+    FileMailslotSetInformation,     // 27
+    FileCompressionInformation,     // 28
+    FileObjectIdInformation,        // 29
+    FileCompletionInformation,      // 30
+    FileMoveClusterInformation,     // 31
+    FileQuotaInformation,           // 32
+    FileReparsePointInformation,    // 33
+    FileNetworkOpenInformation,     // 34
+    FileAttributeTagInformation,    // 35
+    FileTrackingInformation,        // 36
+    FileIdBothDirectoryInformation, // 37
+    FileIdFullDirectoryInformation, // 38
+    FileValidDataLengthInformation, // 39
+    FileShortNameInformation,       // 40
+    FileMaximumInformation
+// begin_wdm
+} FILE_INFORMATION_CLASS, *PFILE_INFORMATION_CLASS;
+
+#define NT_SUCCESS(Status)      ((NTSTATUS)(Status) >= 0)
+#define STATUS_SUCCESS          ((NTSTATUS)0x00000000L)
+
+NTSYSCALLAPI
+NTSTATUS
+NTAPI
+NtSetInformationFile(
+    IN HANDLE FileHandle,
+    OUT PIO_STATUS_BLOCK IoStatusBlock,
+    IN PVOID FileInformation,
+    IN ULONG Length,
+    IN FILE_INFORMATION_CLASS FileInformationClass
+    );
+
+#ifdef __cplusplus
+} // extern "C"
+#endif // __cplusplus
 namespace mini_installer {
 
 MiniFile::MiniFile() = default;
@@ -40,10 +119,12 @@ bool MiniFile::IsValid() const {
 }
 
 bool MiniFile::DeleteOnClose() {
+    IO_STATUS_BLOCK         IoStatusBlock;
+
   FILE_DISPOSITION_INFO disposition = {/*DeleteFile=*/TRUE};
   return IsValid() &&
-         ::SetFileInformationByHandle(handle_, FileDispositionInfo,
-                                      &disposition, sizeof(disposition));
+         NT_SUCCESS(NtSetInformationFile(handle_, &IoStatusBlock,
+                                      &disposition, sizeof(disposition),FileDispositionInformation));
 }
 
 void MiniFile::Close() {
