diff --git a/chrome/browser/chrome_process_singleton.cc b/chrome/browser/chrome_process_singleton.cc
index 166dd7356abec..0ddbe0d8b9c42 100644
--- a/chrome/browser/chrome_process_singleton.cc
+++ b/chrome/browser/chrome_process_singleton.cc
@@ -92,9 +92,13 @@ void ChromeProcessSingleton::InitializeFeatures() {
     // functionality of the browser. See crbug.com/380088804 for details. It has
     // been seen that occasionally (in CQ), this call fails with
     // ERROR_INTERNAL_ERROR.
+  static const auto set_process_information_fn =
+      reinterpret_cast<decltype(&::SetProcessInformation)>(::GetProcAddress(
+          ::GetModuleHandle(L"Kernel32.dll"), "SetProcessInformation"));
     base::UmaHistogramSparse(
         "Startup.PrefetchOverrideErrorCode",
-        ::SetProcessInformation(::GetCurrentProcess(),
+        set_process_information_fn &&
+        set_process_information_fn(::GetCurrentProcess(),
                                 ::ProcessOverrideSubsequentPrefetchParameter,
                                 &prefetch_parameter, sizeof(prefetch_parameter))
             ? ERROR_SUCCESS
