diff --git a/gpu/command_buffer/service/BUILD.gn b/gpu/command_buffer/service/BUILD.gn
index ceefb4a32429d..6cbb60caa4058 100644
--- a/gpu/command_buffer/service/BUILD.gn
+++ b/gpu/command_buffer/service/BUILD.gn
@@ -95,17 +95,6 @@ target(link_target_type, "service_sources") {
   }
 }
 
-if (is_win) {
-  config("service_win_linker_flags") {
-    # This is in a separate config so the flags can be applied to dependents.
-    # ldflags in GN aren't automatically inherited.
-    ldflags = [
-      # Linking with OneCore.lib causes the following dlls to load.
-      "/DELAYLOAD:api-ms-win-core-handle-l1-1-0.dll",
-    ]
-  }
-}
-
 target(link_target_type, "gles2_sources") {
   # External code should depend on this via //gpu/command_buffer/service:gles2
   # above rather than depending on this directly or the component build will
@@ -626,9 +615,7 @@ target(link_target_type, "gles2_sources") {
       "d3d11.lib",
       "d3d12.lib",
       "dxguid.lib",
-      "onecore.lib",
     ]
-    all_dependent_configs = [ ":service_win_linker_flags" ]
   }
 }
 
diff --git a/gpu/command_buffer/service/dxgi_shared_handle_manager.cc b/gpu/command_buffer/service/dxgi_shared_handle_manager.cc
index e6b8b472847af..70a9ac321f096 100644
--- a/gpu/command_buffer/service/dxgi_shared_handle_manager.cc
+++ b/gpu/command_buffer/service/dxgi_shared_handle_manager.cc
@@ -16,6 +16,31 @@ namespace gpu {
 
 namespace {
 
+bool IsSameHandle(HANDLE handle, HANDLE other) {
+  using PFN_COMPARE_OBJECT_HANDLES =
+      BOOL(WINAPI*)(HANDLE hFirstObjectHandle, HANDLE hSecondObjectHandle);
+  static PFN_COMPARE_OBJECT_HANDLES compare_object_handles_fn =
+      []() -> PFN_COMPARE_OBJECT_HANDLES {
+    HMODULE kernelbase_module = ::GetModuleHandle(L"kernelbase.dll");
+    if (!kernelbase_module) {
+      DVLOG(1) << "kernelbase.dll not found";
+      return nullptr;
+    }
+    PFN_COMPARE_OBJECT_HANDLES fn =
+        reinterpret_cast<PFN_COMPARE_OBJECT_HANDLES>(
+            ::GetProcAddress(kernelbase_module, "CompareObjectHandles"));
+    if (!fn)
+      DVLOG(1) << "CompareObjectHandles not found";
+    return fn;
+  }();
+  if (compare_object_handles_fn)
+    return compare_object_handles_fn(handle, other);
+  // CompareObjectHandles isn't available before Windows 10. Return true in that
+  // case since there's no other way to check if the handles refer to the same
+  // D3D11 texture and IsSameHandle is only used as a sanity test.
+  return true;
+}
+
 Microsoft::WRL::ComPtr<ID3D11Texture2D> OpenSharedHandleTexture(
     Microsoft::WRL::ComPtr<ID3D11Device> d3d11_device,
     HANDLE shared_handle) {
@@ -266,7 +291,7 @@ DXGISharedHandleManager::GetOrCreateSharedHandleState(
     DCHECK(state);
     // If there's already a shared handle associated with the token, it should
     // refer to the same D3D11 texture (or kernel object).
-    if (!CompareObjectHandles(shared_handle.Get(), state->GetSharedHandle())) {
+    if (!IsSameHandle(shared_handle.Get(), state->GetSharedHandle())) {
       LOG(ERROR) << "Existing shared handle for token doesn't match";
       return nullptr;
     }
diff --git a/ui/gfx/BUILD.gn b/ui/gfx/BUILD.gn
index 7f04b27b3d9f1..98869537561ba 100644
--- a/ui/gfx/BUILD.gn
+++ b/ui/gfx/BUILD.gn
@@ -57,17 +57,6 @@ component("gfx_skia") {
   defines = [ "GFX_SKIA_IMPLEMENTATION" ]
 }
 
-if (is_win) {
-  config("gfx_win_linker_flags") {
-    # This is in a separate config so the flags can be applied to dependents.
-    # ldflags in GN aren't automatically inherited.
-    ldflags = [
-      # Linking with OneCore.lib causes the following dlls to load.
-      "/DELAYLOAD:api-ms-win-core-handle-l1-1-0.dll",
-    ]
-  }
-}
-
 # Binds together the `gfx` component.
 #
 # See also: `docs/component_build.md`
@@ -398,9 +387,7 @@ component("gfx") {
       "setupapi.lib",
       "dwmapi.lib",
       "dwrite.lib",
-      "onecore.lib",
     ]
-    all_dependent_configs = [ ":gfx_win_linker_flags" ]
     public_deps += [ "//components/crash/core/common" ]
   }
 
diff --git a/ui/gfx/win/d3d_shared_fence.cc b/ui/gfx/win/d3d_shared_fence.cc
index 29ce16d9e83f0..d74a4e793f135 100644
--- a/ui/gfx/win/d3d_shared_fence.cc
+++ b/ui/gfx/win/d3d_shared_fence.cc
@@ -103,7 +103,30 @@ Microsoft::WRL::ComPtr<ID3D11Device> D3DSharedFence::GetD3D11Device() const {
 }
 
 bool D3DSharedFence::IsSameFenceAsHandle(HANDLE shared_handle) const {
-  return CompareObjectHandles(shared_handle_.Get(), shared_handle);
+  using PFN_COMPARE_OBJECT_HANDLES =
+      BOOL(WINAPI*)(HANDLE hFirstObjectHandle, HANDLE hSecondObjectHandle);
+  static PFN_COMPARE_OBJECT_HANDLES compare_object_handles_fn =
+      []() -> PFN_COMPARE_OBJECT_HANDLES {
+    HMODULE kernelbase_module = ::GetModuleHandle(L"kernelbase.dll");
+    if (!kernelbase_module) {
+      DVLOG(1) << "kernelbase.dll not found";
+      return nullptr;
+    }
+    PFN_COMPARE_OBJECT_HANDLES fn =
+        reinterpret_cast<PFN_COMPARE_OBJECT_HANDLES>(
+            ::GetProcAddress(kernelbase_module, "CompareObjectHandles"));
+    if (!fn) {
+      DVLOG(1) << "CompareObjectHandles not found";
+    }
+    return fn;
+  }();
+  if (compare_object_handles_fn) {
+    return compare_object_handles_fn(shared_handle_.Get(), shared_handle);
+  }
+  // CompareObjectHandles isn't available before Windows 10 but we shouldn't be
+  // using fences in that case either.
+  NOTREACHED();
+  return false;
 }
 
 void D3DSharedFence::Update(uint64_t fence_value) {
