diff --git a/base/win/startup_information.cc b/base/win/startup_information.cc
index 372b85e04ffcb..9986674a05941 100644
--- a/base/win/startup_information.cc
+++ b/base/win/startup_information.cc
@@ -6,43 +6,84 @@
 
 #include <utility>
 
+namespace {
+
+typedef BOOL (WINAPI *InitializeProcThreadAttributeListFunction)(
+    LPPROC_THREAD_ATTRIBUTE_LIST attribute_list,
+    DWORD attribute_count,
+    DWORD flags,
+    PSIZE_T size);
+static InitializeProcThreadAttributeListFunction
+    initialize_proc_thread_attribute_list;
+
+typedef BOOL (WINAPI *UpdateProcThreadAttributeFunction)(
+    LPPROC_THREAD_ATTRIBUTE_LIST attribute_list,
+    DWORD flags,
+    DWORD_PTR attribute,
+    PVOID value,
+    SIZE_T size,
+    PVOID previous_value,
+    PSIZE_T return_size);
+static UpdateProcThreadAttributeFunction update_proc_thread_attribute_list;
+
+typedef VOID (WINAPI *DeleteProcThreadAttributeListFunction)(
+    LPPROC_THREAD_ATTRIBUTE_LIST lpAttributeList);
+static DeleteProcThreadAttributeListFunction delete_proc_thread_attribute_list;
+
+}  // namespace
+
 namespace base {
 namespace win {
 
-StartupInformation::StartupInformation() : startup_info_() {
+StartupInformation::StartupInformation() {
+  memset(&startup_info_, 0, sizeof(startup_info_));
   startup_info_.StartupInfo.cb = sizeof(startup_info_);
+
+  // Load the attribute API functions.
+  if (!initialize_proc_thread_attribute_list ||
+      !update_proc_thread_attribute_list ||
+      !delete_proc_thread_attribute_list) {
+    HMODULE module = ::GetModuleHandleW(L"kernel32.dll");
+    initialize_proc_thread_attribute_list =
+        reinterpret_cast<InitializeProcThreadAttributeListFunction>(
+            ::GetProcAddress(module, "InitializeProcThreadAttributeList"));
+    update_proc_thread_attribute_list =
+        reinterpret_cast<UpdateProcThreadAttributeFunction>(
+            ::GetProcAddress(module, "UpdateProcThreadAttribute"));
+    delete_proc_thread_attribute_list =
+        reinterpret_cast<DeleteProcThreadAttributeListFunction>(
+            ::GetProcAddress(module, "DeleteProcThreadAttributeList"));
+  }
 }
 
 StartupInformation::~StartupInformation() {
   if (startup_info_.lpAttributeList) {
-    ::DeleteProcThreadAttributeList(startup_info_.lpAttributeList);
+    delete_proc_thread_attribute_list(startup_info_.lpAttributeList);
+    delete [] reinterpret_cast<BYTE*>(startup_info_.lpAttributeList);
   }
 }
 
 bool StartupInformation::InitializeProcThreadAttributeList(
     DWORD attribute_count) {
   if (startup_info_.StartupInfo.cb != sizeof(startup_info_) ||
-      startup_info_.lpAttributeList) {
+      startup_info_.lpAttributeList)
     return false;
-  }
 
   SIZE_T size = 0;
-  ::InitializeProcThreadAttributeList(nullptr, attribute_count, 0, &size);
+  initialize_proc_thread_attribute_list(NULL, attribute_count, 0, &size);
   if (size == 0) {
     return false;
   }
 
-  auto attribute_list = base::HeapArray<char>::Uninit(size);
-  auto* attribute_list_ptr =
-      reinterpret_cast<LPPROC_THREAD_ATTRIBUTE_LIST>(attribute_list.data());
-  if (!::InitializeProcThreadAttributeList(attribute_list_ptr, attribute_count,
-                                           0, &size)) {
+  startup_info_.lpAttributeList =
+      reinterpret_cast<LPPROC_THREAD_ATTRIBUTE_LIST>(new BYTE[size]);
+  if (!initialize_proc_thread_attribute_list(startup_info_.lpAttributeList,
+                                           attribute_count, 0, &size)) {
+    delete [] reinterpret_cast<BYTE*>(startup_info_.lpAttributeList);
+    startup_info_.lpAttributeList = NULL;
     return false;
   }
 
-  attribute_list_ = std::move(attribute_list);
-  startup_info_.lpAttributeList = attribute_list_ptr;
-
   return true;
 }
 
@@ -51,9 +92,8 @@ bool StartupInformation::UpdateProcThreadAttribute(
   if (!startup_info_.lpAttributeList) {
     return false;
   }
-  return !!::UpdateProcThreadAttribute(startup_info_.lpAttributeList, 0,
-                                       attribute, value, size, nullptr,
-                                       nullptr);
+  return !!update_proc_thread_attribute_list(startup_info_.lpAttributeList, 0,
+                                       attribute, value, size, NULL, NULL);
 }
 
 }  // namespace win
diff --git a/base/win/startup_information.h b/base/win/startup_information.h
index 2d4edaf52113b..5b777baefe13c 100644
--- a/base/win/startup_information.h
+++ b/base/win/startup_information.h
@@ -43,7 +41,6 @@ class BASE_EXPORT StartupInformation {
   }
 
  private:
-  base::HeapArray<char> attribute_list_;
   STARTUPINFOEXW startup_info_;
 };
 
