diff --git a/sandbox/win/src/process_mitigations.cc b/sandbox/win/src/process_mitigations.cc
index 1f2509c7abe28..08a9c89485563 100644
--- a/sandbox/win/src/process_mitigations.cc
+++ b/sandbox/win/src/process_mitigations.cc
@@ -20,6 +20,8 @@
 namespace sandbox {
 
 namespace {
+// API defined in winbase.h >= Vista.
+using SetProcessDEPPolicyFunction = decltype(&SetProcessDEPPolicy);
 
 // API defined in libloaderapi.h >= Win8.
 using SetDefaultDllDirectoriesFunction = decltype(&SetDefaultDllDirectories);
@@ -143,9 +145,15 @@ bool ApplyProcessMitigationsToCurrentProcess(MitigationFlags flags) {
     if (combined_flags & MITIGATION_DEP_NO_ATL_THUNK)
       dep_flags |= PROCESS_DEP_DISABLE_ATL_THUNK_EMULATION;
 
-    if (!::SetProcessDEPPolicy(dep_flags)) {
+    SetProcessDEPPolicyFunction set_process_dep_policy =
+        reinterpret_cast<SetProcessDEPPolicyFunction>(
+            ::GetProcAddress(GetModuleHandleA("kernel32.dll"), "SetProcessDEPPolicy"));
+    if (set_process_dep_policy) {
+      if (!set_process_dep_policy(dep_flags)) {
+        return false;
+      }
+    } else
       return false;
-    }
 
     applied_flags |=
         combined_flags & (MITIGATION_DEP | MITIGATION_DEP_NO_ATL_THUNK);
