diff --git a/sandbox/win/src/target_services.cc b/sandbox/win/src/target_services.cc
index a1895fd2eec1b..cb095ce39530c 100644
--- a/sandbox/win/src/target_services.cc
+++ b/sandbox/win/src/target_services.cc
@@ -79,6 +79,11 @@ bool CloseOpenHandles(bool* is_csrss_connected) {
   return true;
 }
 
+// GetUserDefaultLocaleName is not available on WIN XP.  So we'll
+// load it on-the-fly.
+const wchar_t kKernel32DllName[] = L"kernel32.dll";
+typedef decltype(GetUserDefaultLocaleName)* GetUserDefaultLocaleNameFunction;
+
 // Warm up language subsystems before the sandbox is turned on.
 // Tested on Win8.1 x64:
 // This needs to happen after RevertToSelf() is called, because (at least) in
@@ -91,8 +96,23 @@ bool WarmupWindowsLocales() {
   // warmup all of these functions, but let's not assume that.
   ::GetUserDefaultLangID();
   ::GetUserDefaultLCID();
+  static GetUserDefaultLocaleNameFunction GetUserDefaultLocaleName_func =
+      nullptr;
+  if (!GetUserDefaultLocaleName_func) {
+    HMODULE kernel32_dll = ::GetModuleHandle(kKernel32DllName);
+    if (!kernel32_dll) {
+      return false;
+    }
+    GetUserDefaultLocaleName_func =
+        reinterpret_cast<GetUserDefaultLocaleNameFunction>(
+            GetProcAddress(kernel32_dll, "GetUserDefaultLocaleName"));
+    if (!GetUserDefaultLocaleName_func) {
+      return false;
+    }
+  }
   wchar_t localeName[LOCALE_NAME_MAX_LENGTH] = {};
-  return (0 != ::GetUserDefaultLocaleName(localeName, LOCALE_NAME_MAX_LENGTH));
+  return (0 != GetUserDefaultLocaleName_func(
+                   localeName, LOCALE_NAME_MAX_LENGTH * sizeof(wchar_t)));
 }
 
 bool SetProcessIntegrityLevel(IntegrityLevel integrity_level) {
