diff --git a/components/system_cpu/cpu_probe_win.cc b/components/system_cpu/cpu_probe_win.cc
index 0348393ecf8ae..0f1bd5d69c5a0 100644
--- a/components/system_cpu/cpu_probe_win.cc
+++ b/components/system_cpu/cpu_probe_win.cc
@@ -71,6 +71,40 @@ CpuProbeWin::BlockingTaskRunnerHelper::~BlockingTaskRunnerHelper() {
   DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
 }
 
+PDH_STATUS (WINAPI *g_lpfnPdhAddCounter)(HQUERY, LPCWSTR, DWORD_PTR, HCOUNTER *);
+
+
+
+BOOL
+InitPdhLibrary(
+    HMODULE *phPdhLib    // [OUT] Pointer to the PDH.DLL module.
+)
+{ 
+    BOOL    bRet = FALSE;
+
+    *phPdhLib = LoadLibraryW(L"pdh.dll");
+
+    if(!*phPdhLib)
+    {
+        goto exit;
+    }
+
+    if(!(g_lpfnPdhAddCounter = (PDH_STATUS (WINAPI *)(HQUERY, LPCWSTR, DWORD_PTR, HCOUNTER *))GetProcAddress(*phPdhLib,"PdhAddEnglishCounterW") ))
+    {
+        if(!(g_lpfnPdhAddCounter = (PDH_STATUS (WINAPI *)(HQUERY, LPCWSTR, DWORD_PTR, HCOUNTER *))GetProcAddress(*phPdhLib,"PdhAddCounterW") ))
+        {
+            goto exit;
+        }
+    }
+
+    bRet = TRUE;
+
+exit:
+
+    return bRet;
+}
+
+
 std::optional<CpuSample> CpuProbeWin::BlockingTaskRunnerHelper::Update() {
   DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
 
@@ -78,7 +112,8 @@ absl::optional<PressureSample> CpuProbeWin::BlockingTaskRunnerHelper::Update() {
 
   if (!cpu_query_.is_valid()) {
     cpu_query_ = base::win::ScopedPdhQuery::Create();
-    if (!cpu_query_.is_valid()) {
+    HMODULE         hPdhLib     = NULL;
+    if (!InitPdhLibrary(&hPdhLib) || !cpu_query_.is_valid()) {
       return std::nullopt;
     }
 
@@ -91,7 +126,7 @@ absl::optional<PressureSample> CpuProbeWin::BlockingTaskRunnerHelper::Update() {
     const auto* query_info = is_running_in_vm
                                  ? kHypervisorLogicalProcessorCounter
                                  : kProcessorCounter;
-    pdh_status = PdhAddEnglishCounter(cpu_query_.get(), query_info, NULL,
+    pdh_status = g_lpfnPdhAddCounter(cpu_query_.get(), query_info, NULL,
                                       &cpu_percent_utilization_);
 
     // When Chrome is running under a different hypervisor, we can add the
@@ -100,11 +135,16 @@ absl::optional<PressureSample> CpuProbeWin::BlockingTaskRunnerHelper::Update() {
       pdh_status = PdhCollectQueryData(cpu_query_.get());
       if (pdh_status != ERROR_SUCCESS) {
         query_info = kProcessorCounter;
-        pdh_status = PdhAddEnglishCounter(cpu_query_.get(), query_info, NULL,
+        pdh_status = g_lpfnPdhAddCounter(cpu_query_.get(), query_info, NULL,
                                           &cpu_percent_utilization_);
       }
     }
 
+    if(hPdhLib)
+    {
+        FreeLibrary(hPdhLib);
+    }
+
     if (pdh_status != ERROR_SUCCESS) {
       cpu_query_.reset();
       LOG(ERROR) << "PdhAddEnglishCounter failed for '" << query_info
