diff --git a/crypto/process_bound_string.cc b/crypto/process_bound_string.cc
index 2ca23c54ec2dd..faf6af52e955c 100644
--- a/crypto/process_bound_string.cc
+++ b/crypto/process_bound_string.cc
@@ -35,7 +35,18 @@ size_t MaybeRoundUp(size_t size) {
 bool MaybeEncryptBuffer(base::span<uint8_t> buffer) {
 #if BUILDFLAG(IS_WIN)
   TRACE_EVENT0(kTraceCategory, "ProcessBoundString::EncryptBuffer");
-  if (::CryptProtectMemory(buffer.data(), buffer.size(),
+typedef BOOL (WINAPI *CRYPTPROTECTMEMORY)(LPVOID pData,DWORD cbData,DWORD dwFlags);
+
+    CRYPTPROTECTMEMORY pCryptProtectMemory=NULL;
+    HMODULE hCrypt;
+        hCrypt = GetModuleHandleW(L"Crypt32.dll");
+        if (hCrypt != NULL)
+        {
+          // Available since Vista.
+          pCryptProtectMemory = (CRYPTPROTECTMEMORY)GetProcAddress(hCrypt, "CryptProtectMemory");
+        }
+  if (pCryptProtectMemory)
+  if (pCryptProtectMemory(buffer.data(), buffer.size(),
                            CRYPTPROTECTMEMORY_SAME_PROCESS)) {
     return true;
   }
@@ -45,7 +56,17 @@ bool MaybeEncryptBuffer(base::span<uint8_t> buffer) {
 bool MaybeDecryptBuffer(base::span<uint8_t> buffer) {
 #if BUILDFLAG(IS_WIN)
   TRACE_EVENT0(kTraceCategory, "ProcessBoundString::DecryptBuffer");
-  if (::CryptUnprotectMemory(buffer.data(), buffer.size(),
+typedef BOOL (WINAPI *CRYPTUNPROTECTMEMORY)(LPVOID pData,DWORD cbData,DWORD dwFlags);
+
+    CRYPTUNPROTECTMEMORY pCryptUnprotectMemory=NULL;
+    HMODULE hCrypt;
+        hCrypt = GetModuleHandleW(L"Crypt32.dll");
+        if (hCrypt != NULL)
+        {
+          pCryptUnprotectMemory = (CRYPTUNPROTECTMEMORY)GetProcAddress(hCrypt, "CryptUnprotectMemory");
+        }
+  if (pCryptUnprotectMemory)
+  if (pCryptUnprotectMemory(buffer.data(), buffer.size(),
                              CRYPTPROTECTMEMORY_SAME_PROCESS)) {
     return true;
   }
