diff --git a/components/paint_preview/common/serial_utils.cc b/components/paint_preview/common/serial_utils.cc
index 5579ff789813f..0824334733237 100644
--- a/components/paint_preview/common/serial_utils.cc
+++ b/components/paint_preview/common/serial_utils.cc
@@ -17,14 +17,12 @@
 #include "third_party/skia/include/codec/SkWebpDecoder.h"
 #include "third_party/skia/include/core/SkCanvas.h"
 #include "third_party/skia/include/core/SkData.h"
-#include "third_party/skia/include/core/SkFontMgr.h"
 #include "third_party/skia/include/core/SkImage.h"
 #include "third_party/skia/include/core/SkImageInfo.h"
 #include "third_party/skia/include/core/SkMatrix.h"
 #include "third_party/skia/include/core/SkPictureRecorder.h"
 #include "third_party/skia/include/core/SkStream.h"
 #include "third_party/skia/include/core/SkString.h"
-#include "third_party/skia/include/core/SkTypeface.h"
 #include "third_party/skia/include/private/chromium/Slug.h"
 
 namespace paint_preview {
@@ -110,24 +108,6 @@ sk_sp<SkData> SerializeTypeface(SkTypeface* typeface, void* ctx) {
   return subset_data;
 }
 
-static sk_sp<SkTypeface> DeserializeTypeface(const void* data,
-                                             size_t length,
-                                             void* ctx) {
-  TRACE_EVENT0("paint_preview", "DeserializeTypeface");
-  // TODO(bungeman,kjlubick) This should not be how the Skia deserial proc
-  // works.
-  SkStream* stream = *(reinterpret_cast<SkStream**>(const_cast<void*>(data)));
-  if (length < sizeof(stream)) {
-    return nullptr;
-  }
-  // The default implementation of SkPicture deserialization of SkTypeface
-  // does not use a fallback (system) font manager, but this is necessary
-  // on Android due to the above behavior w/r to system fonts. Thus, we
-  // call the underlying SkTypeface::MakeDeserialize and pass in the
-  // system font manager ourselves.
-  return SkTypeface::MakeDeserialize(stream, skia::DefaultFontMgr());
-}
-
 static bool is_supported_codec(sk_sp<const SkData> data) {
   CHECK(data);
   return SkBmpDecoder::IsBmp(data->data(), data->size()) ||
@@ -316,7 +297,6 @@ SkDeserialProcs MakeDeserialProcs(DeserializationContext* ctx) {
   procs.fPictureProc = DeserializePictureAsRectData;
   procs.fPictureCtx = ctx;
   procs.fImageProc = DeserializeImage;
-  procs.fTypefaceProc = DeserializeTypeface;
   sktext::gpu::Slug::AddDeserialProcs(&procs, nullptr);
   return procs;
 }
@@ -326,7 +306,6 @@ SkDeserialProcs MakeDeserialProcs(LoadedFramesDeserialContext* ctx) {
   procs.fPictureProc = GetPictureFromDeserialContext;
   procs.fPictureCtx = ctx;
   procs.fImageProc = DeserializeImage;
-  procs.fTypefaceProc = DeserializeTypeface;
   return procs;
 }
 
diff --git a/components/services/font/public/cpp/font_loader.h b/components/services/font/public/cpp/font_loader.h
index 0c5e548e7d5f7..1e7f6110da296 100644
--- a/components/services/font/public/cpp/font_loader.h
+++ b/components/services/font/public/cpp/font_loader.h
@@ -46,8 +46,7 @@ class FontLoader : public SkFontConfigInterface,
                        SkFontStyle* out_style) override;
   SkStreamAsset* openStream(const FontIdentity& identity) override
       LOCKS_EXCLUDED(mapped_font_files_lock_);
-  sk_sp<SkTypeface> makeTypeface(const FontIdentity& identity,
-                                 sk_sp<SkFontMgr> mgr) override
+  sk_sp<SkTypeface> makeTypeface(const FontIdentity& identity) override
       LOCKS_EXCLUDED(typeface_cache_lock_);
 
   // Additional cross-thread accessible methods below.
diff --git a/skia/BUILD.gn b/skia/BUILD.gn
index 5f9362162e48d..7d6900c33a344 100644
--- a/skia/BUILD.gn
+++ b/skia/BUILD.gn
@@ -84,8 +84,6 @@ config("skia_config") {
     "SK_DISABLE_LEGACY_VK_GRDIRECTCONTEXT_FACTORIES",
     "SK_DEFAULT_TYPEFACE_IS_EMPTY",
     "SK_DISABLE_LEGACY_DEFAULT_TYPEFACE",
-    "SK_DISABLE_LEGACY_FONTMGR_REFDEFAULT",
-    "SK_DISABLE_LEGACY_FONTMGR_FACTORY",
 
     # Let Skia and Skia clients know whether to use `libpng`-based or
     # Rust-based PNG decoder and encoder.  This mimics what is done in
