diff --git a/build/config/win/BUILD.gn b/build/config/win/BUILD.gn
index c51ad032acd3f..e7a194de59010 100644
--- a/build/config/win/BUILD.gn
+++ b/build/config/win/BUILD.gn
@@ -95,7 +95,7 @@ config("compiler") {
     cflags += [ "/D__WRL_ENABLE_FUNCTION_STATICS__" ]
 
     # Tell clang which version of MSVC to emulate.
-    cflags += [ "-fmsc-version=1934" ]
+    cflags += [ "-fmsc-version=1916" ]
 
     if (is_component_build) {
       cflags += [
