diff --git a/base/debug/invalid_access_win.cc b/base/debug/invalid_access_win.cc
index d6a96de167b28..a68c18ae4b54d 100644
--- a/base/debug/invalid_access_win.cc
+++ b/base/debug/invalid_access_win.cc
@@ -19,8 +19,13 @@ namespace {
 void CreateSyntheticHeapCorruption() {
   EXCEPTION_RECORD record = {};
   record.ExceptionCode = STATUS_HEAP_CORRUPTION;
-  RaiseFailFastException(&record, nullptr,
-                         FAIL_FAST_GENERATE_EXCEPTION_ADDRESS);
+        HMODULE k32handle = GetModuleHandleW(L"kernelbase.dll");
+        auto pfnRaiseFailFastException =
+            (decltype(RaiseFailFastException)*)GetProcAddress(k32handle, "RaiseFailFastException");
+        if (pfnRaiseFailFastException)
+        {
+            pfnRaiseFailFastException(&record, nullptr, FAIL_FAST_GENERATE_EXCEPTION_ADDRESS);
+        }
 }
 
 }  // namespace
