diff --git a/chrome/app/chromium_strings.grd b/chrome/app/chromium_strings.grd
index 9bb1ba1..40e9cb1 100644
--- a/chrome/app/chromium_strings.grd
+++ b/chrome/app/chromium_strings.grd
@@ -420,7 +420,7 @@ Chromium is unable to recover your settings.
          Can not install the same Chromium version that is currently running. Please close Chromium and try again.
          </message>
         <message name="IDS_INSTALL_OS_NOT_SUPPORTED" desc="Error displayed if OS is not supported">
-          Chromium requires Windows 7 or higher.
+          Chromium requires Windows Vista or Windows XP with SP2 or higher.
         </message>
         <message name="IDS_INSTALL_OS_ERROR" desc="Error displayed when any Windows API call fails and we do not have more specific information.">
           An operating system error occurred during installation. Please download Chromium again.
diff --git a/chrome/app/google_chrome_strings.grd b/chrome/app/google_chrome_strings.grd
index a9edbad..d0ef9d3 100644
--- a/chrome/app/google_chrome_strings.grd
+++ b/chrome/app/google_chrome_strings.grd
@@ -421,7 +421,7 @@ Google Chrome is unable to recover your settings.
           Can not install the same Google Chrome version that is currently running. Please close Google Chrome and try again.
         </message>
         <message name="IDS_INSTALL_OS_NOT_SUPPORTED" desc="Error displayed if OS is not supported">
-          Google Chrome requires Windows 7 or higher.
+          Google Chrome requires Windows Vista or Windows XP with SP2 or higher.
         </message>
         <message name="IDS_INSTALL_OS_ERROR" desc="Error displayed when any Windows API call fails and we do not have more specific information.">
           An operating system error occurred during installation. Please download Google Chrome again.
diff --git a/chrome/installer/setup/setup_main.cc b/chrome/installer/setup/setup_main.cc
index 085fe65..7a4022c 100644
--- a/chrome/installer/setup/setup_main.cc
+++ b/chrome/installer/setup/setup_main.cc
@@ -1733,10 +1733,10 @@ int WINAPI wWinMain(HINSTANCE instance, HINSTANCE prev_instance,
     persistent_histogram_storage.Disable();
   }
 
-  // Check to make sure current system is Win7 or later. If not, log
+  // Check to make sure current system is WinXP or later. If not, log
   // error message and get out.
   if (!InstallUtil::IsOSSupported()) {
-    LOG(ERROR) << "Chrome only supports Windows 7 or later.";
+    LOG(ERROR) << "Chrome only supports Windows XP or later.";
     installer_state.WriteInstallerResult(installer::OS_NOT_SUPPORTED,
                                          IDS_INSTALL_OS_NOT_SUPPORTED_BASE,
                                          nullptr);
diff --git a/chrome/installer/util/install_util.cc b/chrome/installer/util/install_util.cc
index b07d1a4..e894cbb 100644
--- a/chrome/installer/util/install_util.cc
+++ b/chrome/installer/util/install_util.cc
@@ -271,10 +271,12 @@ void InstallUtil::GetCriticalUpdateVersion(BrowserDistribution* dist,
 }
 
 bool InstallUtil::IsOSSupported() {
-  // We do not support anything prior to Windows 7.
+  // We do not support Win2K or older, or XP without service pack 2.
   VLOG(1) << base::SysInfo::OperatingSystemName() << ' '
           << base::SysInfo::OperatingSystemVersion();
-  return base::win::GetVersion() >= base::win::Version::WIN7;
+  base::win::Version version = base::win::GetVersion();
+  return (version > base::win::Version::XP) ||
+      (version == base::win::Version::XP);
 }
 
 void InstallUtil::AddInstallerResultItems(bool system_install,
