diff --git a/base/process/process_metrics_win.cc b/base/process/process_metrics_win.cc
index 3a0400b75aeee..91a4c2eb9d646 100644
--- a/base/process/process_metrics_win.cc
+++ b/base/process/process_metrics_win.cc
@@ -226,7 +226,7 @@ ProcessMetrics::ProcessMetrics(ProcessHandle process) {
   HANDLE duplicate_handle = INVALID_HANDLE_VALUE;
   BOOL result = ::DuplicateHandle(::GetCurrentProcess(), process,
                                   ::GetCurrentProcess(), &duplicate_handle,
-                                  PROCESS_QUERY_LIMITED_INFORMATION, FALSE, 0);
+                                  PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, 0);
   if (!result) {
     // Even with PROCESS_QUERY_LIMITED_INFORMATION, DuplicateHandle can fail
     // with ERROR_ACCESS_DENIED. And it's always possible to run out of handles.
