diff --git a/sandbox/policy/win/sandbox_win.cc b/sandbox/policy/win/sandbox_win.cc
index fc71a3cb8dc38..7e0fa0d1b456d 100644
--- a/sandbox/policy/win/sandbox_win.cc
+++ b/sandbox/policy/win/sandbox_win.cc
@@ -941,11 +941,6 @@ sandbox::ResultCode SandboxWin::StartSandboxedProcess(
     return result;
   }
 
-  if (sandbox_type == Sandbox::kGpu) {
-    config->SetLockdownDefaultDacl();
-    config->AddRestrictingRandomSid();
-  }
-
   CHECK_EQ(AddGenericConfig(config), SBOX_ALL_OK);
 
   std::string appcontainer_id;
